src/rustcore/src/admin.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2019 Twitter, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use pelikan::core::admin::AdminHandler;
+use pelikan::protocol::{PartialParseError, Protocol, QuitRequest, Serializable};
+
+use std::cell::RefCell;
+use std::io::{Result, Write};
+use std::net::SocketAddr;
+use std::rc::Rc;
+use std::time::{Duration, Instant};
+
+use tokio::net::{TcpListener, TcpStream};
+use tokio::prelude::*;
+use tokio::runtime::current_thread::spawn;
+use tokio::timer::Interval;
+
+// Uncomment once tokio updates to 0.2.0-alpha.7
+// use tokio::signal::CtrlC;
+// use futures::select;
+
+use ccommon::buf::OwnedBuf;
+use ccommon_sys::{buf, buf_sock_borrow, buf_sock_return};
+
+type Request<H> = <<H as AdminHandler>::Protocol as Protocol>::Request;
+type Response<H> = <<H as AdminHandler>::Protocol as Protocol>::Response;
+
+#[inline(never)]
+fn assert_buf_valid(buf: &mut OwnedBuf) {
+    unsafe {
+        let ptr: *mut *mut buf = std::mem::transmute(buf);
+
+        let mut value = true;
+        value = value && (**ptr).rpos as usize <= (**ptr).end as usize;
+        value = value && (**ptr).wpos as usize <= (**ptr).end as usize;
+        value = value && (**ptr).rpos as usize <= (**ptr).wpos as usize;
+
+        if !value {
+            panic!();
+        }
+    }
+}
+
+/// Process a single request stream
+async fn admin_tcp_stream_handler<H, S>(handler: Rc<RefCell<H>>, mut stream: S)
+where
+    H: AdminHandler + 'static,
+    S: AsyncWrite + AsyncRead + Unpin,
+    <H::Protocol as Protocol>::Request: QuitRequest,
+{
+    let mut sock = unsafe { buf_sock_borrow() };
+
+    let (mut rbuf, mut wbuf) = unsafe {
+        (
+            OwnedBuf::from_raw((*sock).wbuf),
+            OwnedBuf::from_raw((*sock).rbuf),
+        )
+    };
+
+    let mut tmpbuf = [0u8; 1024];
```

Comment:
why using `tmpbuf` instead of directly reading into `rbuf`?

Reply:
I didn't want to have to worry about uninitialized memory here. However, once twitter/ccommon#226 is ported over it will be possible to read directly into the buffer.

Reply:
This has been done now